pyiceberg/table/__init__.py

```diff
@@ -1884,8 +1884,9 @@ def to_arrow_batch_reader(self) -> pa.RecordBatchReader:
 
         from pyiceberg.io.pyarrow import project_batches, schema_to_pyarrow
 
+        target_schema = schema_to_pyarrow(self.projection())
```

Comment:
My preference would be to let Arrow decide. For Polars it is different because they are also the query engine. Casting the types will recompute the buffers, consuming additional memory/CPU, which I would rather avoid.

For the table, we first materialize all the batches in memory, so if one of them is large, it will automatically upcast, otherwise, it will keep the small types.

Reply:
My knowledge on Parquet data to Arrow buffer conversion is less versed, so please do check me if I am not making much sense 🙂 

But are we actually casting the types on read?

We make a decision on whether we are choosing to read with large or small types when instantiating the [fragment scanner](https://github.com/apache/iceberg-python/blob/main/pyiceberg/io/pyarrow.py#L1052C20-L1052C54), which loads the parquet data into the Arrow buffers. The `schema_to_pyarrow()` calls to `pa.Table` or `pa.RecordBatchReader` or in `to_requested_schema` following that all represent the Table schema in the consistent (large or small) format which shouldn't result in any additional casting and reassigning of buffers.

I think the only time we are casting the types is on write, where we may want to downcast it for forward compatibility. It looks like we have to choose a schema to use on write anyways, because using a schema for the ParquetWriter that isn't consistent with the schema within the dataframe results in an [exception](https://github.com/apache/iceberg-python/pull/902/files#r1669524329).